File: backend/src/strategy/schemas.py
"""strategy 도메인 Pydantic V2 스키마."""

from __future__ import annotations

from typing import Literal
from uuid import UUID

from pydantic import AwareDatetime, BaseModel, ConfigDict, Field, field_validator

from src.strategy.models import ParseStatus, PineVersion
from src.strategy.trading_sessions import validate_session_names


class CreateStrategyRequest(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=2000)
    pine_source: str = Field(min_length=1)
    timeframe: str | None = Field(default=None, max_length=16)
    symbol: str | None = Field(default=None, max_length=32)
    tags: list[str] = Field(default_factory=list)
    # Sprint 7d: empty list = 24h. Subset of {"asia","london","ny"}.
    trading_sessions: list[str] = Field(default_factory=list)

    @field_validator("trading_sessions")
    @classmethod
    def _validate_sessions(cls, v: list[str]) -> list[str]:
        return validate_session_names(v)


class UpdateStrategyRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=2000)
    pine_source: str | None = Field(default=None, min_length=1)
    timeframe: str | None = Field(default=None, max_length=16)
    symbol: str | None = Field(default=None, max_length=32)
    tags: list[str] | None = None
    trading_sessions: list[str] | None = None
    is_archived: bool | None = None

    @field_validator("trading_sessions")
    @classmethod
    def _validate_sessions(cls, v: list[str] | None) -> list[str] | None:
        return validate_session_names(v) if v is not None else None


class ParseRequest(BaseModel):
    pine_source: str = Field(min_length=1)


class ParseError(BaseModel):
    code: str
    message: str
    line: int | None = None


# Sprint 29 Slice B: 미지원 호출 상세 응답 schema (line + workaround + category 포함)
class UnsupportedCallResponse(BaseModel):
    """미지원 호출 상세 — DrFXGOD reject 응답에 line 번호 + 우회 안내 포함."""

    name: str
    line: int
    col: int | None = None
    workaround: str | None = None
    category: Literal["drawing", "data", "syntax", "math", "other"]


# Sprint 29 Slice B: CoverageReport 전체 응답 schema (parse-preview API 확장 대상)
class CoverageReportResponse(BaseModel):
    """coverage analyzer 결과 전체 응답 — backward-compat + Slice B 확장 필드."""

    is_runnable: bool
    used_functions: list[str]
    used_attributes: list[str]
    unsupported_functions: list[str]  # 기존 backward-compat
    unsupported_attributes: list[str]  # 기존 backward-compat
    unsupported_calls: list[UnsupportedCallResponse] = Field(
        default_factory=list
    )  # Sprint 29 Slice B
    dogfood_only_warning: str | None = None  # Sprint 29 Slice A


class ParsePreviewResponse(BaseModel):
    status: ParseStatus
    pine_version: PineVersion
    warnings: list[str] = Field(default_factory=list)
    errors: list[ParseError] = Field(default_factory=list)
    entry_count: int = 0
    exit_count: int = 0
    # Sprint 7b ISSUE-004: UI 파싱 결과 탭 '감지된 지표/전략 콜' 섹션 렌더링을 위해
    # parser supported_feature_report["functions_used"]를 응답에 노출.
    functions_used: list[str] = Field(default_factory=list)
    # Sprint Y1 (B+D): pre-flight coverage analyzer — 미지원 built-in 명시.
    # `unsupported_builtins` 가 비어있을 때만 backtest 실행 가능 (CLAUDE.md Golden Rule).
    unsupported_builtins: list[str] = Field(default_factory=list)
    # Sprint 29 Slice B: line 번호 + workaround 포함 상세 응답
    unsupported_calls: list[UnsupportedCallResponse] = Field(default_factory=list)
    # Sprint 29 Slice A: heikinashi Trust Layer 위반 경고
    dogfood_only_warning: str | None = None
    # 실행 가능 여부 (FE 가 backtest 버튼 비활성화 + 안내 표시 결정에 사용)
    is_runnable: bool = True


class StrategySettings(BaseModel):
    """Sprint 26 — Live Signal Auto-Trading 의 trading params.

    schema_version: 향후 schema 변경 시 backward compat 식별 (P3 #2).
    leverage / margin_mode: Bybit Futures dispatch 분기 의무 (Sprint 22 BL-091).
    position_size_pct: 가용 잔고 대비 포지션 크기 (0-100, 100 = all-in).

    extra="forbid" 로 잘못된 필드 422 reject — codex G.0 P2 #4 malformed JSONB 방어.
    """

    model_config = ConfigDict(extra="forbid")

    schema_version: int = 1
    leverage: int = Field(ge=1, le=125)
    margin_mode: Literal["cross", "isolated"]
    position_size_pct: float = Field(gt=0, le=100)


def validate_strategy_settings(
    raw: dict[str, object] | None,
) -> StrategySettings | None:
    """JSONB → StrategySettings parse. None = unset (no-op). 실패 시 ValidationError.

    Sprint 26 codex G.0 P2 #4 — read path 에서 모든 strategy.settings 사용 전 검증.
    """
    if raw is None:
        return None
    return StrategySettings.model_validate(raw)


class UpdateStrategySettingsRequest(BaseModel):
    """PUT /strategies/{id}/settings request body. StrategySettings 와 동일 schema."""

    model_config = ConfigDict(extra="forbid")

    schema_version: int = 1
    leverage: int = Field(ge=1, le=125)
    margin_mode: Literal["cross", "isolated"]
    position_size_pct: float = Field(gt=0, le=100)


class StrategyListItem(BaseModel):
    """목록 DTO — pine_source/description 제외."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    pine_version: PineVersion
    parse_status: ParseStatus
    parse_errors: list[dict[str, object]] | None = None
    timeframe: str | None = None
    symbol: str | None = None
    tags: list[str] = Field(default_factory=list)
    trading_sessions: list[str] = Field(default_factory=list)
    settings: dict[str, object] | None = None  # Sprint 26 (P3 #1)
    is_archived: bool
    created_at: AwareDatetime
    updated_at: AwareDatetime


class StrategyResponse(BaseModel):
    """상세 DTO — 전 필드 포함."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    description: str | None
    pine_source: str
    pine_version: PineVersion
    parse_status: ParseStatus
    parse_errors: list[dict[str, object]] | None
    timeframe: str | None
    symbol: str | None
    tags: list[str] = Field(default_factory=list)
    trading_sessions: list[str] = Field(default_factory=list)
    settings: dict[str, object] | None = None  # Sprint 26 (P3 #1)
    is_archived: bool
    created_at: AwareDatetime
    updated_at: AwareDatetime


class StrategyCreateResponse(StrategyResponse):
    """create 응답 전용. 신규 발급된 webhook_secret plaintext 1회 포함.

    Sprint 13 Phase A.1.4: Strategy 생성 시 atomic auto-issue 된 secret 의 plaintext 를
    1회만 응답에 포함. GET / list 응답에서는 절대 사용 금지 — frontend 는 sessionStorage
    캐시 (TTL 30분) 로 재사용.
    """

    webhook_secret: str | None = None


class StrategyListResponse(BaseModel):
    items: list[StrategyListItem]
    total: int
    page: int
    limit: int
    total_pages: int

File: frontend/src/features/strategy/schemas.ts
// Sprint 7c: Strategy domain Zod 스키마 — Backend API 응답/요청 런타임 검증.
// 타입은 z.infer로 단일 소스화 (schema-first 원칙).

import { z } from "zod/v4";

export const ParseStatusSchema = z.enum(["ok", "unsupported", "error"]);
export type ParseStatus = z.infer<typeof ParseStatusSchema>;

export const PineVersionSchema = z.enum(["v4", "v5"]);
export type PineVersion = z.infer<typeof PineVersionSchema>;

export const ParseErrorSchema = z.object({
  code: z.string(),
  message: z.string(),
  line: z.number().int().nullable(),
});
export type ParseError = z.infer<typeof ParseErrorSchema>;

export const ParsePreviewResponseSchema = z.object({
  status: ParseStatusSchema,
  pine_version: PineVersionSchema,
  warnings: z.array(z.string()).default([]),
  errors: z.array(ParseErrorSchema).default([]),
  entry_count: z.number().int().default(0),
  exit_count: z.number().int().default(0),
  // Sprint 7b ISSUE-004: BE ParseOutcome.supported_feature_report["functions_used"] 반영.
  functions_used: z.array(z.string()).default([]),
  // Sprint Y1: pre-flight coverage analyzer — 미지원 built-in 명시 (whack-a-mole 종식)
  unsupported_builtins: z.array(z.string()).default([]),
  is_runnable: z.boolean().default(true),
});
export type ParsePreviewResponse = z.infer<typeof ParsePreviewResponseSchema>;

export const TradingSessionSchema = z.enum(["asia", "london", "ny"]);
export type TradingSession = z.infer<typeof TradingSessionSchema>;

// Sprint 27 BL-137 — trading settings (Live Signal Auto-Trading 의 leverage/margin/size).
// Backend StrategySettings (backend/src/strategy/schemas.py:72-87) 와 동일 spec.
export const MarginModeSchema = z.enum(["cross", "isolated"]);
export type MarginMode = z.infer<typeof MarginModeSchema>;

// codex G.2 P1 #1 — BE extra='forbid' 정합. FE 가 unknown key 통과시키면
// 백엔드에서 422, 또는 FE schema 가 silent strip → BE 와 mismatch.
export const StrategySettingsSchema = z
  .object({
    schema_version: z.number().int().default(1),
    leverage: z.number().int().min(1).max(125),
    margin_mode: MarginModeSchema,
    position_size_pct: z.number().gt(0).max(100),
  })
  .strict();
export type StrategySettings = z.infer<typeof StrategySettingsSchema>;

export const StrategyResponseSchema = z.object({
  id: z.uuid(),
  name: z.string(),
  description: z.string().nullable(),
  pine_source: z.string(),
  pine_version: PineVersionSchema,
  parse_status: ParseStatusSchema,
  parse_errors: z.array(z.record(z.string(), z.unknown())).nullable(),
  timeframe: z.string().nullable(),
  symbol: z.string().nullable(),
  tags: z.array(z.string()).default([]),
  trading_sessions: z.array(z.string()).default([]),
  // Sprint 27 BL-137 — settings JSONB. null = unset (Live Session 시작 차단).
  settings: StrategySettingsSchema.nullable().optional(),
  is_archived: z.boolean(),
  created_at: z.iso.datetime({ offset: true }),
  updated_at: z.iso.datetime({ offset: true }),
});
export type StrategyResponse = z.infer<typeof StrategyResponseSchema>;

// Sprint 13 Phase A.1.4: Strategy create 응답에만 webhook_secret plaintext 1회 포함.
// GET / list 응답은 StrategyResponse 유지 — webhook_secret 노출 X.
export const StrategyCreateResponseSchema = StrategyResponseSchema.extend({
  webhook_secret: z.string().nullable().optional(),
});
export type StrategyCreateResponse = z.infer<typeof StrategyCreateResponseSchema>;

// Sprint 13 Phase A.2: rotate-webhook-secret 응답 (Sprint 6 broken bug fix 후).
export const WebhookRotateResponseSchema = z.object({
  secret: z.string(),
  webhook_url: z.string(),
});
export type WebhookRotateResponse = z.infer<typeof WebhookRotateResponseSchema>;

export const StrategyListItemSchema = StrategyResponseSchema.omit({
  pine_source: true,
  description: true,
});
export type StrategyListItem = z.infer<typeof StrategyListItemSchema>;

export const StrategyListResponseSchema = z.object({
  items: z.array(StrategyListItemSchema),
  total: z.number().int(),
  page: z.number().int(),
  limit: z.number().int(),
  total_pages: z.number().int(),
});
export type StrategyListResponse = z.infer<typeof StrategyListResponseSchema>;

export const CreateStrategyRequestSchema = z.object({
  name: z.string().min(1).max(120),
  description: z.string().max(2000).nullable().optional(),
  pine_source: z.string().min(1),
  timeframe: z.string().max(16).nullable().optional(),
  symbol: z.string().max(32).nullable().optional(),
  tags: z.array(z.string()).default([]),
});
export type CreateStrategyRequest = z.infer<typeof CreateStrategyRequestSchema>;

export const UpdateStrategyRequestSchema = z.object({
  name: z.string().min(1).max(120).optional(),
  description: z.string().max(2000).nullable().optional(),
  pine_source: z.string().min(1).optional(),
  timeframe: z.string().max(16).nullable().optional(),
  symbol: z.string().max(32).nullable().optional(),
  tags: z.array(z.string()).optional(),
  trading_sessions: z.array(z.string()).optional(),
  is_archived: z.boolean().optional(),
});
export type UpdateStrategyRequest = z.infer<typeof UpdateStrategyRequestSchema>;

// Sprint 27 BL-137 — PUT /strategies/{id}/settings request body. Backend
// UpdateStrategySettingsRequest (extra="forbid") 와 동일 spec.
export const UpdateStrategySettingsRequestSchema = StrategySettingsSchema;
export type UpdateStrategySettingsRequest = z.infer<
  typeof UpdateStrategySettingsRequestSchema
>;

export const StrategyListQuerySchema = z.object({
  limit: z.number().int().min(1).max(100).default(20),
  offset: z.number().int().min(0).default(0),
  parse_status: ParseStatusSchema.optional(),
  is_archived: z.boolean().default(false),
});
export type StrategyListQuery = z.infer<typeof StrategyListQuerySchema>;
